feat #12164: Add events for Creaking heart and Eyeblossom blocks

Guilherme-Corvelo opened the pull request:
- Introduced CreakingHeartTransformEvent and EyeblossomTransformEvent to the API, enabling plugins to monitor and control state transitions of Creaking Heart and Eyeblossom blocks.
- Ensured all state changes for Creaking Heart (AWAKE, DORMANT, UPROOTED) and Eyeblossom (OPEN, CLOSED) are routed through their respective events, preventing untracked transitions.
- Added CreakingHeartState enum to the API for stable plugin-side state referencing, decoupling plugin logic from NMS internals.
- Implemented CreakingHeartProtectorChangeEvent to allow plugins to intercept protector (Creaking mob) spawn and removal.
- Updated server-side logic to convert NMS state enums to API enums before firing events, ensuring API consistency.

These changes enhance plugin interoperability, event-driven extensibility, and long-term maintainability of custom block behaviors.

Guilherme-Corvelo committed:
feat #12164: Add events for Creaking heart and Eyeblossom blocks

- Introduced CreakingHeartTransformEvent and
  EyeblossomTransformEvent to the API, enabling plugins to
  monitor and control state transitions of Creaking Heart and
  Eyeblossom blocks.
- Ensured all state changes for Creaking Heart (AWAKE, DORMANT,
  UPROOTED) and Eyeblossom (OPEN, CLOSED) are routed through
  their respective events, preventing untracked transitions.
- Added CreakingHeartState enum to the API for stable
  plugin-side state referencing, decoupling plugin logic from
  NMS internals.
- Implemented CreakingHeartProtectorChangeEvent to allow
  plugins to intercept protector (Creaking mob) spawn and
  removal.
- Updated server-side logic to convert NMS state enums to API
  enums before firing events, ensuring API consistency.

These changes enhance plugin interoperability, event-driven
extensibility, and long-term maintainability of custom block
behaviors.

Signed-off-by: Guilherme Corvelo <guilherme.corvelo@tecnico.ulisboa.pt>

Co-authored-by: Afonso Mendonça Jacinto <afonso.mendonca.jacinto@tecnico.ulisboa.pt>

Strokkur424 commented on paper-api/src/main/java/org/bukkit/block/CreakingHeartState.java:
```diff
@@ -0,0 +1,29 @@
+    public String toString() {
+        return this.name;
+    }
+
```
I don't think there is any specific reason to include a constructor for the lowercased name for the `toString()` method

Lulu13022002 replied:
in fact this enum already exists under CreakingHeart$State not sure why it has been duplicated here

AfonsoMendoncaJacinto replied:
The point was really just to decouple NMS internals from API logic.

Brokkonaut replied:
It is api: https://jd.papermc.io/paper/1.21.5/org/bukkit/block/data/type/CreakingHeart.State.html

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartProtectorChangeEvent.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * The type of action being performed on the protector.
+     */
+    public enum Action { SPAWN, REMOVE }
```
Instead of having a tiny Action enum, I think it would be better to just have a field `spawning`, which, if `false`, indicates that the protector was removed instead.

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartProtectorChangeEvent.java:
```diff
@@ -0,0 +1,85 @@
+    @NotNull
+    public Action getAction() {
+        return action;
+ }
```
The formatting seems to be a bit broken

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,114 @@
+        @NotNull BlockState oldState,
+        @NotNull BlockState newState,
+        @NotNull CreakingHeartState oldLogicalState,
+        @NotNull CreakingHeartState newLogicalState
```
This is true for all classes, not only this one. But instead of adding the `@NotNull` annotation everywhere, you should instead just annotate the class as `@NullMarked` (imported from `jspecify`). All of the jetbrains `@NotNull` and `@Nullable` should also be replaced with their jspecify equivalents, since that is what we are aiming to switch to.

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,114 @@
+    private final BlockState oldState;
+    private final BlockState newState;
+    private final CreakingHeartState oldLogicalState;
+    private final CreakingHeartState newLogicalState;
```
I kinda dislike these being fields called 'logical states'. Maybe instead just call them `(old|new)CreatingHeartState`?

Strokkur424 commented on paper-server/patches/sources/net/minecraft/world/level/block/CreakingHeartBlock.java.patch:
```diff
@@ -0,0 +1,49 @@
+     }
+ 
+     private static BlockState updateState(BlockState state, Level level, BlockPos pos) {
++        net.minecraft.world.level.block.state.properties.CreakingHeartState oldState = state.getValue(STATE);
```
Please don't forget the `// Paper start` comments

Strokkur424 commented on paper-server/patches/sources/net/minecraft/world/level/block/CreakingHeartBlock.java.patch:
```diff
@@ -0,0 +1,49 @@
++
++        if (oldState != newState && level instanceof net.minecraft.server.level.ServerLevel serverLevel) {
++
++            // Paper Begin: Fire CreakingHeartTransformEvent before changing block
```
This should be `// Paper start` instead of `// Paper Begin`, and also as mentioned already, above the whole block of your changes

Strokkur424 commented on paper-server/patches/sources/net/minecraft/world/level/block/CreakingHeartBlock.java.patch:
```diff
@@ -0,0 +1,49 @@
++            bukkitNewState.setBlockData(org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(newNmsState));
++            bukkitNewState.setType(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(state.getBlock()));
++
++            // Conversion NMS -> API
```
There is no need for this comment here

Strokkur424 commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,92 @@
+     }
+ 
+     private static BlockState updateCreakingState(Level level, BlockState state, BlockPos pos, CreakingHeartBlockEntity creakingHeart) {
++        CreakingHeartState oldState = state.getValue(CreakingHeartBlock.STATE);
```
Again here, please don't forget the `// Paper start` comments

Strokkur424 commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,92 @@
+         } else {
+             Creaking creaking = optional.get();
++
++            // Paper Begin: Fire CreakingHeartProtectorChangeEvent (SPAWN)
```
`// Paper Begin` -> `// Paper start`

Strokkur424 commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,92 @@
+     public void removeProtector(@Nullable DamageSource damageSource) {
+         if (this.getCreakingProtector().orElse(null) instanceof Creaking creaking) {
++
++            // Paper Begin: Fire CreakingHeartProtectorChangeEvent (REMOVE)
```
Here as well

Guilherme-Corvelo committed:
feat #12164: Add events for Creaking heart and Eyeblossom blocks
- Introduced CreakingHeartTransformEvent and
  EyeblossomTransformEvent to the API, enabling plugins to
  monitor and control state transitions of Creaking Heart and
  Eyeblossom blocks both in the wild and in flower pots.
- Ensured all state changes for Creaking Heart (AWAKE, DORMANT,
  UPROOTED) and Eyeblossom (OPEN, CLOSED) are routed through
  their respective events, preventing untracked transitions.
- Implemented CreakingHeartProtectorChangeEvent to allow
  plugins to intercept protector (Creaking mob) spawn and
  removal.
- Updated server-side logic to convert NMS state enums to API
  enums before firing events, ensuring API consistency.
These changes enhance plugin interoperability, event-driven
extensibility, and long-term maintainability of custom block
behaviors.

Signed-off-by: Guilherme Corvelo <guilherme.corvelo@tecnico.ulisboa.pt>

Co-authored-by: Afonso Mendonça Jacinto <afonso.mendonca.jacinto@tecnico.ulisboa.pt>

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartProtectorChangeEvent.java:
```diff
@@ -0,0 +1,84 @@
+     * @param protector the protector entity (may be null for REMOVE if not available)
+     * @param spawn the indicator of whether the mob is spawned or not
+     */
+    public CreakingHeartProtectorChangeEvent(Block block, @Nullable LivingEntity protector, boolean spawn) {
```
Constructors in the API are usually annotated with `@ApiStatus.Internal`. You should do so here as well

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartProtectorChangeEvent.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    
```
You seem to have misplaced the javadocs here?

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartProtectorChangeEvent.java:
```diff
@@ -0,0 +1,84 @@
+     * @return the protector entity, or null if not available
+     */
+    @Nullable
+    public LivingEntity getProtector() { return protector; }
```
Please make sure to not put methods on a single line like this

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartProtectorChangeEvent.java:
```diff
@@ -0,0 +1,84 @@
+     *
+     * @return the boolean
+     */
+    public boolean getIsSpawned() {
```
Rename the method to `isSpawned` and remove the second line of the JavaDocs - that is unnecessary information.

I also kind of dislike the wording here. Maybe the following JD would be better?
```java
/**
 * Get whether the protector is being spawned.
 *
 * @return whether the protector is being spawned
 */
public boolean isSpawned() {
```

AfonsoMendoncaJacinto replied:
Seems right. We will change our solution to that. Thanks

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,111 @@
+     * @param newState the new state of the block after transformation
+     * @param oldState the previous state of the block before transformation
+     */
+    public CreakingHeartTransformEvent(
```
Here as well, please add the `@ApiStatus.Internal` annotation

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return the old logical state
+     */
+    public State getoldCreakingHeartState() { 
```
Rename to `getOldCreakingHeartState`

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return the new logical state
+     */
+    public State getnewCreakingHeartState() { 
```
And here rename to `getNewCreakingHeartState`

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    /**
+     * Gets the previous logical state of the Creaking Heart.
```
Not a big deal, but following English grammar, you would usually write `creaking heart` in all lowercase

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/CreakingHeartTransformEvent.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    /**
+     * Gets the new logical state of the Creaking Heart.
```
Same here

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/EyeblossomTransformEvent.java:
```diff
@@ -0,0 +1,67 @@
+ * <li>Eyeblossom closing during the day.
+ * </ul>
+ * <p>
+ * If this event is cancelled, the block will not transform.
```
This line is unnecessary information - this is true for all cancelable events

Strokkur424 commented on paper-api/src/main/java/org/bukkit/event/block/EyeblossomTransformEvent.java:
```diff
@@ -0,0 +1,67 @@
+     * @param block the block being transformed
+     * @param newState the new state of the block after transformation
+     */
+    public EyeblossomTransformEvent(final Block block,final BlockState newState) {
```
`@ApiStatus.Internal`

Guilherme-Corvelo committed:
feat #12164: Add events for Creaking heart and Eyeblossom blocks
- Introduced CreakingHeartTransformEvent and
  EyeblossomTransformEvent to the API, enabling plugins to
  monitor and control state transitions of Creaking Heart and
  Eyeblossom blocks both in the wild and in flower pots.
- Ensured all state changes for Creaking Heart (AWAKE, DORMANT,
  UPROOTED) and Eyeblossom (OPEN, CLOSED) are routed through
  their respective events, preventing untracked transitions.
- Implemented CreakingHeartProtectorChangeEvent to allow
  plugins to intercept protector (Creaking mob) spawn and
  removal.
- Updated server-side logic to convert NMS state enums to API
  enums before firing events, ensuring API consistency.
These changes enhance plugin interoperability, event-driven
extensibility, and long-term maintainability of custom block
behaviors.

Signed-off-by: Guilherme Corvelo <guilherme.corvelo@tecnico.ulisboa.pt>

Co-authored-by: Afonso Mendonça Jacinto <afonso.mendonca.jacinto@tecnico.ulisboa.pt>

Strokkur424 reviewed (approved):
I didn't test the changes, but they look fine code-wise
